integration/ipam.go

```diff
@@ -0,0 +1,118 @@
+// +build k8srequired
+
+package integration
+
+import (
+	"net"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/giantswarm/ipam"
+	"github.com/giantswarm/microerror"
+)
+
+const (
+	defaultCIDRMask   = 24
+	totalBitsLength   = 32
+	privateSubnetMask = 25
+	publicSubnetMask  = 25
+	networkCIDR       = "10.1.0.0/16"
+)
+
+type awsVPC struct {
+	NetworkCIDR       string
+	PrivateSubnetCIDR string
+	PublicSubnetCIDR  string
+}
+
+func newAWSVPCBlock(client AWSClient) (awsVPC, error) {
+	cidrMask := net.CIDRMask(defaultCIDRMask, totalBitsLength)
+	existingSubnets, err := listSubnets(client)
+	if err != nil {
+		return awsVPC{}, microerror.Mask(err)
+	}
+	cidr, err := newSubnet(cidrMask, existingSubnets)
+	if err != nil {
+		return awsVPC{}, microerror.Mask(err)
+	}
+
+	// Configure private and public subnets as subnets of the new subnet.
+	privateSubnetMask := net.CIDRMask(privateSubnetMask, totalBitsLength)
+	privateSubnetCIDR, err := ipam.Free(cidr, privateSubnetMask, nil)
```

Comment:
When we use ipam magic may this affect peering? I'd like to hear @JosephSalisbury opinion here.

Reply:
All this comes from kubernetesd https://github.com/giantswarm/kubernetesd/blob/master/service/creator/aws.go#L108, the only part specific to this project is querying the list of VPCs from AWS instead of using microstorage

Reply:
Doesn't that mean possible races?

Reply:
Yes, we don't have access to the local list of networks occupied by guest clusters on gauss that kubernetesd has, only to what AWS knows about them.

Reply:
Do you think this is ok?

Reply:
IMO It would be better if all the interested parties would access the same source of truth. Currently, if we create a local example cluster (not e2e), kubernetesd doesn't know anything about the newly occupied CIDR and could mistakenly try to allocate the same block. IMO having a local copy of a remote state is a potential source of this kind of problems.

Reply:
Scratch my above, this will cause cluster creation on gauss / ginger to fail when an integration test is running. 

host cluster network: 10.0.0.0/16, clusters have /24 mask

- integration tests grabs free network (10.0.0.0/24)
- kubernetesd allocates network (also, 10.0.0.0/24)
- bork

i think if we do this it needs to at least be in a seperate account from gauss/ginger

Reply:
+1 separate account would work. I think we have potential to push too many changes to afford such race.

Reply:
Thanks Joe, makes perfect sense. I think having a separate account/host cluster for integration testing of this repo is just too much, maybe if we set up a staging environment would make more sense but only for this one is too much IMO. For this repo I think that just having in place the required network resources, dedicated only to automated integration testing, would be enough, is that correct @JosephSalisbury? That would be a VPC (with a CIDR not overlapping with anything else) two route tables and maybe two worker subnets? Probably a NAT gateway too.

Reply:
Ok! This seems to be working fine:

* I've created a new "host" VPC (and all the additional required network resources) with CIDR 10.11.0.0/16 with aws-terraform (this would be equivalent to gauss' 10.0.0.0/16, not overlapping with it of course).
* Then, in e2e the base CIDR for the guest clusters is 10.12.0.0/16 (equivalent to gauss' guests in 10.1.0.0/16, also not overlapping), and the subnets are created in the 10.12.0.0/24 block.

The changes to use these network blocks are already pushed and this build was run using them https://circleci.com/gh/giantswarm/aws-operator/2314

@JosephSalisbury @kopiczko PTAL, we now need only to care about possible collisions among e2e builds, but if the probability of coincident guest cluster creations and e2e runs was low I think that 2 e2e tests requesting a CIDR block at exactly the same time is even lower, also retriggering the failing test would solve the issue. Let me know WDYT.

Reply:
That works for me.